xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,393 @@
+		return
+	}
+	gsb.cc.UpdateAddresses(sc, addrs)
+}
+
+// SwitchTo gracefully switches to the new balancer. This function must be
+// called synchronously alongside the rest of the balancer.Balancer methods this
+// Graceful Switch Balancer implements.
+func (gsb *gracefulSwitchBalancer) SwitchTo(builder balancer.Builder) error {
+	gsb.mu.Lock()
+	if gsb.closed {
+		gsb.mu.Unlock()
+		return errBalancerClosed
+	}
+	bw := &balancerWrapper{
+		gsb:       gsb,
+		lastState: defaultState,
+		subconns:  make(map[balancer.SubConn]bool),
+	}
+	var balToClose balancer.Balancer
+	if gsb.balancerCurrent == nil {
+		gsb.balancerCurrent = bw
+	} else {
+		// Clean up resources here that are from a previous pending lb.
+		if gsb.balancerPending != nil {
+			for sc := range gsb.balancerPending.subconns {
+				gsb.cc.RemoveSubConn(sc)
+			}
+			balToClose = gsb.balancerPending
+		}
+		gsb.balancerPending = bw
+	}
+	gsb.mu.Unlock()
+	if balToClose != nil {
+		balToClose.Close()
+	}
+
+	newBalancer := builder.Build(bw, gsb.bOpts)
+	if newBalancer == nil {
+		if gsb.balancerPending != nil {
+			gsb.balancerPending = nil
+		} else {
+			gsb.balancerCurrent = nil
+		}
+		return balancer.ErrBadResolverState
+	}
+
+	// This write doesn't need to take gsb.mu because this field never gets read
+	// or written to on any calls from the current or pending. Calls from grpc
+	// to this balancer are guaranteed to be called synchronously, so this
+	// bw.Balancer field will never be forwarded to until this SwitchTo()
+	// function returns.
+	bw.Balancer = newBalancer
+	return nil
+}
+
+// Returns nil if the graceful switch balancer is closed.
+func (gsb *gracefulSwitchBalancer) latestBalancer() balancer.Balancer {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if gsb.closed {
+		return nil
+	}
+	if gsb.balancerPending != nil {
+		return gsb.balancerPending
+	}
+	return gsb.balancerCurrent
+}
+
+// UpdateClientConnState simply forwards the update to one of it's children.
+func (gsb *gracefulSwitchBalancer) UpdateClientConnState(state balancer.ClientConnState) error {
+	balToUpdate := gsb.latestBalancer()
+	if balToUpdate == nil {
+		return errBalancerClosed
+	}
+
+	// Forwarding the update outside of holding gsb.mu prevents deadlock
+	// scenarios, as the update can induce an inline callback which requires
+	// gsb.mu. However, the current balancer can be closed at any time after
+	// giving up gsb.mu. The pending balancer cannot be closed, as we are
+	// guaranteed grpc will call the balancer API synchronously, and for the
+	// pending to also be closed would require another update from grpc. The
+	// current being closed is not a problem for this UpdateClientConnState()
+	// call, as in the scenario that the current balancer can be closed (current
+	// + pending balancers are both populated), this update will always be
+	// forwarded to the pending. Thus, there is a guarantee that this will not
+	// break the balancer API of the balancer by updating after closing.
```

Comment:
This is still really long.  But the explanation of why code does not exist is tricky.  "Through long and careful planning, we do not need to hold `gsb.mu` when calling into child balancers."  :)

Seriously, though, maybe the mutexes can be documented:

```go
// mu protects the following fields and all fields within balancerCurrent and balancerPending.
// mu does not need to be held when calling into the child balancers, as all calls into these
// children happen only as a direct result of a call into the gracefulSwitchBalancer, which are
// also guaranteed to be synchronous
mu sync.Mutex
```

Then delete this and below?

Reply:
> I called the balancer updates outside of the mutex to prevent deadlocks, not because of the synchronous guarantee

You did it to ignore the deadlocks, but it is only _possible_ to do it that way because of the synchronous guarantee.  And because we do need `currentMu`, there should be a small change:

```go
// mu protects the following fields and all fields within balancerCurrent and balancerPending.
// mu does not need to be held when calling into the child balancers, as all calls into these
// children happen only as a direct result of a call into the gracefulSwitchBalancer, which are
// also guaranteed to be synchronous.  There is one exception: an UpdateState call from a
// child balancer can lead to calling Close() on that child.  To prevent that racing with an
// UpdateSubConnState from the channel, we hold currentMu during Close and
// UpdateSubConnState calls.
mu sync.Mutex
```

This could be simplified to:

```go
// Perform this call without gsb.mu to prevent deadlocks if the child calls back into the channel.
// The latest balancer can never be closed during a call from the channel, even without gsb.mu held.
```

Reply:
Rewrote the mutex comment a little for specificity into what can be closed (your comment didn't wrap pending updateState causing current to be deleted). Simplified UpdateClientConnState and ResolverError comments.